[mlir:python] Improve `mlir_(attribute|type|value)_subclass` for `nanobind`s `stubgen`

ingomueller-net opened the pull request:
This PR makes several improvements to the stubs that are created by `mlir_(attribute|type|value)_subclass`.

First, the PR sets the `__module__` attribute of the classes generated by the nanobind adaptors for attributes, types, and values (via `mlir_(attribute|type|value)_subclass`). By default, the `__module__` property is set to `importlib._bootstrap`, which isn't where we want the new class to live. The new logic sets the property to the name of the module provided as `scope` instead. This also makes nanobind's `stubgen` generate stubs for those classes properly, which ignores classes whose `__module__` does not correspond to the module it is generating stubs for. This resolves #127518.

Second, the PR overwrites the function signatures generated by `stubgen` to a format that uses the desired type names (e.g., `mlir.ir.Attribute` instead of `MlirAttribute`).

Finally, the PR piggy-backs some minor doc and style improvements to `PythonAdaptors.h`.

ingomueller-net committed:
[mlir:python] Set `__module__` classes generated by nanobind adaptors.

This PR sets the `__module__` attribute of the classes generated by the
nanobind adaptors for attributes, types, and values (via
`mlir_(attribute|type|value)_subclass`). By default, the `__module__`
property is set to `importlib._bootstrap`, which isn't where we want the
new class to live. The new logic sets the property to the name of the
module provided as `scope` instead. This also makes nanobind's `stubgen`
generate stubs for those classes properly, which ignores classes whose
`__module__` does not correspond to the module it is generating stubs
for. This resolves #127518.

Signed-off-by: Ingo Müller <ingomueller@google.com>

ingomueller-net commented:
@hawkinsp: You might also be interested.

hawkinsp reviewed (approved):
LGTM, for what that's worth.

ingomueller-net committed:
Fix signature of `isinstance` bindings.

This commit overwrites the signature of the `isinstance` member
functions created by `mlir_(attribute|type|value)_subclass` by adding a
`nanobind::signature` argument to the `def` call that creates the
functions. This improves the typing information that `stubgen` creates
for such classes.

Signed-off-by: Ingo Müller <ingomueller@google.com>

ingomueller-net committed:
Minor unrelated doc and formatting improvements (NFC).

This commit fixes three occurrences of the same grammar issue in some
comment and adds a clang-format suppression for a specific ordering of
includes.

Signed-off-by: Ingo Müller <ingomueller@google.com>

ingomueller-net commented:
Thanks, @hawkinsp, for the review! I realized that some more tweaks are required to generate fully correct stubs and expanded the scope of this PR to include them.

makslevental reviewed (approved):
LGTM. I don't think we have tests for type hints but I trust you verified this works.
